### PR TITLE
3864 add notification purpose

### PR DESCRIPTION
**Pull request by kpethtel**
**Story card:** [ch3864](https://app.clubhouse.io/simpledotorg/story/3864/add-purpose-to-notifications)

## Because

Communication types like, "missed_visit_sms_reminder" conflate the mode of communication with the purpose of the communication. So we're simplifying `Communication#communication_type` to `sms` and `whatsapp` and moving the purpose to the notifications model.

### CODE REVIEW

File: app/models/notification.rb
Comment (kpethtel) [inline app/models/notification.rb:27]: TBH, it's pretty clunky as either a prefix or suffix.

File: db/migrate/20210620144826_add_purpose_to_notifications.rb
Comment (kpethtel) [inline db/migrate/20210620144826_add_purpose_to_notifications.rb:4]: All notifications currently in the wild should be from the covid medication reminders.

File: db/migrate/20210620144826_add_purpose_to_notifications.rb
Comment (kpethtel) [inline db/migrate/20210620144826_add_purpose_to_notifications.rb:1]: It may have been smarter to do the migration in its own PR rather than wrapping it up with a bunch of other code changes.

File: db/migrate/20210620144826_add_purpose_to_notifications.rb
Comment (harimohanraj89) [reply db/migrate/20210620144826_add_purpose_to_notifications.rb:4]: We could be more robust by grabbing `Notification.where(subject: nil)` only. We can also set `Notification.where.not(subject: nil)` to `missed_visit_reminder` as well for thoroughness.

File: db/migrate/20210620144826_add_purpose_to_notifications.rb
Comment (harimohanraj89) [reply db/migrate/20210620144826_add_purpose_to_notifications.rb:1]: It's simple enough, I think we can include it here. It also avoids the window of time where the column exists, but the code's not setting it.

File: app/jobs/appointment_notification/worker.rb
Comment (harimohanraj89) [inline app/jobs/appointment_notification/worker.rb:45]: Should we keep the old types around for a while (`when "whatsapp", "missed_visit_whatsapp_reminder"`) to handle any jobs that are executed before the migration switches everyone's comms type? It would be a "transitional backwards compatibility", similar to the communication types hash below - https://github.com/simpledotorg/simple-server/pull/2613/files#diff-8bef04c7770083f39d473b82194bccd100e857238d53212f8a79d72731dbdcebR12-R14

**Review by harimohanraj89 — CHANGES_REQUESTED** [review]
This is neat! Initial feedback below. Could you also give us an overview (in the PR desc) of what this whole ecosystem looks like after this change? What are the different parameters for any given message, etc?

File: app/jobs/appointment_notification/worker.rb
Comment (kpethtel) [reply app/jobs/appointment_notification/worker.rb:45]: The reason I kept the types in the enum is because I want the records to continue to be valid and update-able. I didn't keep them here because there shouldn't be a case where we want to send any of the notifications that currently exist. However... I've been wrong many times before, and it's probably not a bad idea to keep there here to be safe. I'll re-add and add a note in the migration ticket to remove the references.

File: app/jobs/appointment_notification/worker.rb
Comment (rsanheim) [inline app/jobs/appointment_notification/worker.rb:64]: This looks good, but can you create a custom Exception class to raise?  Maybe `UnknownCommunicationType`.

File: db/migrate/20210620144826_add_purpose_to_notifications.rb
Comment (kpethtel) [reply db/migrate/20210620144826_add_purpose_to_notifications.rb:1]: This is definitely the more correct thing thing to do. I didn't do it because I believe the only notifications that exist on prod are for medication reminders. I believe @prabhanshuguptagit confirmed that last week. I've implemented your suggestion here, but  before we run this migration on prod we should confirm that all notifications without subjects are medication reminders and that either 1) there are no notifications with subjects or 2) all notifications with subjects are appointment reminders. You can use the Notification#message attribute to distinguish them. I've already confirmed that there are no notifications in the db in Bangladesh.

File: app/models/notification.rb
Comment (rsanheim) [inline app/models/notification.rb:31]: All of these strings for the `purpose` can be symbols...I'll just comment on this one but the others in this PR can also get changed over.  Since these are basically enums a symbol better represents that.

File: app/models/notification.rb
Comment (rsanheim) [reply app/models/notification.rb:25]: Maybe just leave suffix and prefix off? These are long strings as they are, don't think they need the prefix/suffix.

File: db/migrate/20210620144826_add_purpose_to_notifications.rb
Comment (rsanheim) [inline db/migrate/20210620144826_add_purpose_to_notifications.rb:9]: (blocking) You can use [update_all](https://apidock.com/rails/v4.0.2/ActiveRecord/Relation/update_all) to do all these updates in two SQL statements.  No reason to iterate over every record, especially in a migration where we want things to complete quickly.

**Review by rsanheim — CHANGES_REQUESTED** [review]
I like how this clarifies things.  Left some feedback, one item blocking.

File: config/locales/api/am_ET.yml
Comment (kpethtel) [inline config/locales/api/am_ET.yml:15]: Hopefully this change doesn't cause confusion. There are no actual communication records for OTP  requests, but SMS is pretty obviously wrong, since it contains whatsapp translations and will probably also end up with Imo translations.

File: app/jobs/appointment_notification/worker.rb
Comment (harimohanraj89) [reply app/jobs/appointment_notification/worker.rb]: Yeah let's leave the backwards compatibility in here for a day, and then rip it out.

File: db/migrate/20210620144826_add_purpose_to_notifications.rb
Comment (harimohanraj89) [reply db/migrate/20210620144826_add_purpose_to_notifications.rb:1]: I think this reply was for the previous comment - https://github.com/simpledotorg/simple-server/pull/2613#discussion_r656844355

The less we have to keep in mind while shipping this, the better. Looks like we've updated the migration to operate on covid reminders and missed visit reminders separately. Nice 💯

**Comment by harimohanraj89** [comment]
Heads up @kpethtel - There's one more translation file that needs updating now (once you've merged in the latest master) - `api/sid_ET.yml`. It was added in #2601 .

The rest of my feedback has been addressed. I'm leaving my blocking review here so you don't miss the Sidama file 😄 Feel free to dismiss my review once it's fixed!

File: app/models/notification.rb
Comment (kpethtel) [reply app/models/notification.rb:25]: That's a good idea. Thanks. Hadn't occurred to me to just leave it off.

File: db/migrate/20210620144826_add_purpose_to_notifications.rb
Comment (kpethtel) [reply db/migrate/20210620144826_add_purpose_to_notifications.rb:1]: 🤦 
Funny thing is I noticed your comment persisted after I made the change and thought you had left it at the top of the relevant code block so it wouldn't disappear when I changed the line in question.

File: app/jobs/appointment_notification/worker.rb
Comment (kpethtel) [reply app/jobs/appointment_notification/worker.rb:64]: Will do.

File: db/migrate/20210620144826_add_purpose_to_notifications.rb
Comment (kpethtel) [reply db/migrate/20210620144826_add_purpose_to_notifications.rb:9]: Does that not lock the db? I thought that was the purpose in batching.

**Comment by kpethtel** [comment]
> Heads up @kpethtel - There's one more translation file that needs updating now (once you've merged in the latest master) - `api/sid_ET.yml`. It was added in #2601 .
> 
> The rest of my feedback has been addressed. I'm leaving my blocking review here so you don't miss the Sidama file 😄 Feel free to dismiss my review once it's fixed!

Thanks for the heads up. I'll take care of it.

File: app/models/notification.rb
Comment (kpethtel) [reply app/models/notification.rb:31]: I do prefer symbols...

File: app/models/facility.rb
Comment (kpethtel) [inline app/models/facility.rb:249]: Not directly related to the PR. We added a new language and I'm adding it to our locales map.

File: app/models/notification.rb
Comment (kpethtel) [reply app/models/notification.rb:31]: Changing to symbols didn't work as expected. I think you can use symbols for enums like `purpose[:missed_visit_reminder]` but it seems the actual values are strings.

**Review by harimohanraj89 — APPROVED** [review]

File: db/migrate/20210620144826_add_purpose_to_notifications.rb
Comment (harimohanraj89) [reply db/migrate/20210620144826_add_purpose_to_notifications.rb]: We have ~200k notifications? I think it's preferable to lock and update at once than running 200k queries. A good middle ground would be to grab the records in batches of, say, 10,000, instead of one by one. 

Rough pseudocode: `Thing.where(condition).find_in_batches(10000) { |batch| batch.update_all }`

Comment (kpethtel) [reply db/migrate/20210620144826_add_purpose_to_notifications.rb]: Yeah, that seems reasonable. `find_each` uses `find_in_batches(1_000)`, so the only difference is batch size. I think the criticism Rob was initially making is that I should've been updating in batches rather than individually and I misread him as, "don't use find_each". I think using find_each with update_all satisfies his original request.

**Pull request merged**
